### PR TITLE
WIP Create friends ACL and get rid of ACCESS_FRIENDS

### LINKED ISSUE #3391: Friends access does not inherit properly (Trac #3391)
_Original ticket http://trac.elgg.org/ticket/3391 on 41310617-08-02 by trac user slyhne, assigned to unknown._

(New summary by @mrclay: When someone comments on/likes a post with ACCESS_FRIENDS, the ACCESS_FRIENDS value is inherited, but not the owner. As a result, the users who can see the comment/like is a _different set of users_ than who can see the original post.)

Elgg version: 1.7

I have just experienced some strange access things regarding annotations on a entity (in this case a blog).

User A creates a blog, setting access to friends only.

User B comments on that blog. Comment gets an access of friends only

User C wants to read comments, but is not friend with User B, so User C can only read the blog, not see comment given by User B. User C gives a comment, which then can only be read by User A and and User C, not User B.

I can see from the code, that annotations allways inherit the access_id of the entity they belong to. That must give some issues when using anything else than public!

A fix could be to allways set annotations to public access, and then make sure they are only shown to those who have access to the entity they belong to.

### PR DESCRIPTION
**Pull request by jdalsem**
Fixes #3391, #5038
Refs #8073, #3872

### CODE REVIEW

File: engine/classes/ElggEntity.php [hunk 1]
```diff
@@ -897,6 +897,28 @@ public function getOwnedAccessCollections($options = []) {
...
+	 * Returns the first ACL owned by the entity with a given subtype
+	 *
+	 * @param string $subtype subtype of the ACL
+	 *
+	 * @return \ElggAccessCollection|false
+	 *
+	 * @since 3.0
+	 */
+	public function getOwnedAccessCollection($subtype) {
+		if (!is_string($subtype)) {
```
Comment (jeabakker) [inline engine/classes/ElggEntity.php]: `|| $subtype === ''`

File: engine/classes/ElggEntity.php [hunk 2]
```diff
@@ -1301,7 +1323,11 @@ protected function create() {
 		$container_guid = (int) $container_guid;
 
 		if ($access_id == ACCESS_DEFAULT) {
-			throw new \InvalidParameterException('ACCESS_DEFAULT is not a valid access level. See its documentation in elgglib.h');
+			throw new \InvalidParameterException('ACCESS_DEFAULT is not a valid access level. See its documentation in elgglib.php');
```
Comment (jeabakker) [inline engine/classes/ElggEntity.php]: not `elgglib.php` but in `constants.php` ....

File: engine/lib/access.php [hunk 3]
```diff
@@ -506,6 +506,106 @@ function access_init() {
 }
 
 /**
+ * Creates a Friends ACL for a user
+ *
+ * @param \Elgg\Event $event event
+ *
+ * @return void
+ */
```
Comment (jeabakker) [inline engine/lib/access.php]: `@since` and `@internal`

File: engine/lib/access.php [hunk 4]
```diff
@@ -531,5 +631,11 @@ function access_test($hook, $type, $value, $params) {
 	// and the user is logged in so it can start caching
 	$events->registerHandler('ready', 'system', 'access_init');
 
+	// friends ACL events
+	$events->registerHandler('create', 'user', 'access_friends_acl_create');
```
Comment (jeabakker) [inline engine/lib/access.php]: check if `create:after` is available

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/lib/access.php (hunk 4 above)
Comment (jdalsem) [reply engine/lib/access.php]: not available

**Comment by jdalsem** [comment]
grr... going to redo this

**Pull request closed without merging**